### PR TITLE
Client Registry introduced and refactored code to handle Deployments/Rollouts

### PR DESCRIPTION
**Pull request by deefreak**
…rollouts

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -126,6 +127,7 @@ type Config struct {
 		EventScaleUpBufferPeriodInHours int    `yaml:"eventScaleUpBufferPeriodInHours"`
 		CustomEventDataConfigMapName    string `yaml:"customEventDataConfigMapName"`
 	} `yaml:"eventCallIntegration"`
+	EnableRollout *bool `yaml:"enableRollout"`
```
Comment (bharathguvvala) [inline cmd/main.go:130]: we can rename this to EnabledArgoRolloutsSupport

File: cmd/main.go [hunk 2]
```diff
@@ -243,7 +245,14 @@ func main() {
...
 		metricsTransformer = append(metricsTransformer, outlierInterpolatorTransformer)
 	}
+	clientsRegistryBuilder := registry.NewDeploymentClientRegistryBuilder().
+		WithK8sClient(mgr.GetClient()).
+		WithCustomDeploymentClient(registry.DeploymentGVK)
 
+	if *config.EnableRollout {
+		clientsRegistryBuilder = clientsRegistryBuilder.WithCustomDeploymentClient(registry.RolloutGVK)
+	}
+	clientsRegistry := clientsRegistryBuilder.Build()
```
Comment (bharathguvvala) [inline cmd/main.go:255]: you can prefix this with deploymentClientRegistry

File: pkg/reco/reco.go [hunk 3]
```diff
@@ -309,108 +307,42 @@ func (c *CpuUtilizationBasedRecommender) calculateSavings(maxReplicas int, simul
...
 func (c *CpuUtilizationBasedRecommender) getContainerCPULimitsSum(namespace, objectKind, objectName string) (float64,
 	error) {
-	var obj client.Object
-	switch objectKind {
-	case "Deployment":
-		obj = &appsv1.Deployment{}
-	case "Rollout":
-		obj = &rolloutv1alpha1.Rollout{}
-	default:
+	obj, err := c.clientsRegistry.GetObjectClient(objectKind)
```
Comment (bharathguvvala) [inline pkg/reco/reco.go:310]: rename obj to deploymentClient

File: pkg/controller/deployment_trigger_controller.go [hunk 4]
```diff
@@ -53,31 +52,16 @@ func (r *DeploymentTriggerController) Reconcile(ctx context.Context,
...
-
-	if !errors.IsNotFound(err) {
-		logger.Error(err, "Failed to get Deployment. Requeue the request")
-		return ctrl.Result{RequeueAfter: 1 * time.Second}, err
+	for _, obj := range r.ClientsRegistry.Clients {
+		object, err := obj.GetObject(request.Namespace, request.Name)
+		if err == nil {
+			return ctrl.Result{}, r.requeuePolicyRecommendation(ctx, object, r.Scheme, logger)
+		}
+		if !errors.IsNotFound(err) {
```
Comment (deefreak) [inline pkg/controller/deployment_trigger_controller.go:60]: Move this outside of the for loop.

**Review by bharathguvvala — APPROVED** [review]

**Pull request merged**